catalog/app/components/Preview/loaders/IFrame/BrowsableLoader.tsx

```diff
@@ -0,0 +1,195 @@
+      const res = await refreshSession({ id, ttl })
+      if (res.error) throw res.error
+      if (!res.data) throw new Error('No data')
+      const r = res.data.browsingSessionRefresh
+      switch (r.__typename) {
+        case 'BrowsingSession':
+          return r
+        case 'OperationError':
+          throw new Error(r.message)
+        case 'InvalidInput':
+          throw new Error(
+            r.errors
+              .map(({ message, path }) => `{message: ${message}, path: ${path} }`)
+              .join('\n'),
+          )
+        default:
+          throw r
+      }
+    },
+    [refreshSession],
+  )
+}
+
+function useDisposeSession() {
+  const [, disposeSession] = urql.useMutation(DISPOSE_BROWSING_SESSION)
+  return React.useCallback(
+    (id?: string) => (id ? disposeSession({ id }) : null),
+    [disposeSession],
+  )
+}
+
+interface FileHandle extends LogicalKeyResolver.S3SummarizeHandle {
+  packageHandle: PackageHandle
+}
+
+interface BrowsableLoaderProps {
+  children: (result: $TSFixMe) => React.ReactNode
+  handle: FileHandle
+}
+
+function useSession(handle: FileHandle) {
+  const [error, setError] = React.useState<Error | unknown | null>(null)
+  const [loading, setLoading] = React.useState(true)
+  const [session, setSession] = React.useState<Session | null>(null)
+  const [key, setKey] = React.useState(0)
+  const retry = React.useCallback(() => {
+    setError(null)
+    setLoading(true)
+    setSession(null)
+    setKey(R.inc)
+  }, [])
+
+  const createSession = useCreateSession()
+  const disposeSession = useDisposeSession()
+  const refreshSession = useRefreshSession()
+
+  const scope = PackageUri.stringify(handle.packageHandle)
+
+  React.useEffect(() => {
+    let ignore = false
+    let sessionClosure: Session | null = null
+
+    async function requestSession() {
+      try {
+        const s = await createSession(scope, SESSION_TTL)
+        if (ignore) return
+        sessionClosure = s
+        setSession(s)
+      } catch (e) {
+        const err = isBucketNotBrowsableError(e as ErrorLike)
+          ? PreviewError.Forbidden()
+          : PreviewError.Unexpected({ retry })
+        setError(err)
+        log.error(e)
+      }
+      setLoading(false)
+    }
+
+    requestSession()
+
+    return () => {
+      disposeSession(sessionClosure?.id)
+      ignore = true
+    }
+  }, [createSession, disposeSession, key, retry, scope])
+
+  React.useEffect(() => {
+    if (!session) return
+    // Refresh when there is 20% left of time to session's end
+    const delay = (session.expires.getTime() - Date.now()) * 0.8
+    const timer = setTimeout(async () => {
+      try {
+        const s = await refreshSession(session.id, SESSION_TTL)
+        setSession(s)
+      } catch (e) {
+        const err = isSessionNotFoundError(e as ErrorLike)
+          ? PreviewError.Expired({ retry })
+          : PreviewError.Unexpected({ retry })
+        setError(err)
+        log.error(e)
```

Comment:
Yes, but I believe it would be more correct to call sentry inside log.error, not in component directly.
We can implement this ourselves or use transport system like one offered by `winston`

Reply:
i explored that route -- tried to integrate sentry and loglevel -- and it seemed quite inflexible. a lot of unnecessary stuff ends up in sentry. so i believe we should explicitly decide which errors to send to sentry